popup.js

```diff
@@ -3,12 +3,101 @@ document.addEventListener('DOMContentLoaded', function() {
 });
 
 function click() {
-  chrome.tabs.executeScript(null, {file: "content_script.js"});
+  var messages = []
+  var results;
+  chrome.runtime.onMessage.addListener(function(message, sender, sendResponse) {
+    messages.push(message);
+    if (results && messages.length == results.length) {
+      completeProcess(messages);
+    }
+  });
+
+  chrome.tabs.executeScript(null, {file: 'content_script.js', allFrames: true}, function(response) { // TODO: FIGURE OUT WHY NOT ALL IFRAMES ARE GETTING CONTENT SCRIPTS INJECTED
+    results = response;
+    if (messages.length == results.length) {
+      completeProcess(messages);
+    }
+  });
+}
+
+/**
+ * Takes all the responses from the injected content scripts and creates the HTML file for download
+ * 
+ * @param {Array<Object>} messages The response from all of the injected content scripts.
+ */
+function completeProcess(messages) {
+  var html = outputHTMLString(messages);
+  var file = new Blob([html], {type: 'text/html'});
+  var url = URL.createObjectURL(file);
+
+  var a = document.getElementById('download');
+  a.href = url;
+  a.download = "webpage.html";
+  a.innerHTML = "Download";
+}
+
+/**
+ * Converts the responses from the injected content scripts into a string representing the HTML.
+ * 
+ * @param {Array<Object>} messages The response from all of the injected content scripts.
+ * @return {string} The resulting HTML
+ */
+function outputHTMLString(messages) {
+  var rootIndex = 0;
+  for (var i = 1; i < messages.length; i++) {
+    rootIndex = messages[i].frameIndex === '0' ? i : rootIndex;
+  }
+  fillRemainingHoles(messages, rootIndex, 0);
+  return messages[rootIndex].html.join('');
+}
+
+/**
+ * Fills all of the gaps in messages[i].html
+ *
+ * @param {Array<Object>} messages The response from all of the injected content scripts
+ * @param {number} i The index of messages to use
+ * @param {number} depth How many parent iframes messages[i] has
+ */
+function fillRemainingHoles(messages, i, depth) {
+  var html = messages[i].html;
+  var quotes = getQuotes(depth);
+  var frameHoles = messages[i].frameHoles;
+  for (var index in frameHoles) {
+    if (frameHoles.hasOwnProperty(index)) {
```

Comment:
I was going off of this stack overflow post: http://stackoverflow.com/questions/684672/how-do-i-loop-through-or-enumerate-a-javascript-object

The top answer says it's important.  Is this not the case?